rhio/examples/topics.rs

```diff
@@ -0,0 +1,69 @@
+use std::time::Duration;
+
+use anyhow::Result;
+use p2panda_core::PrivateKey;
+use rhio::config::Config;
+use rhio::messages::{Message, MessageContext};
+use rhio::node::Node;
+use rhio::topic_id::TopicId;
+
+/// The only message type in our chat app
+type ChatMessage = String;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    let private_key = PrivateKey::new();
+    let mut config = Config::default();
+
+    // Add a topic to the node configuration
+    let chat_topic_id = TopicId::from_str("my_chat");
+    config.topics.push(chat_topic_id);
```

Comment:
Isn't the subscription-at-runtime feature already provided by `subscribe(topic_id)` in the `network` module of `p2panda-net`?

`// Subscribes to a topic and establishes a bi-directional stream from which we can read and write to.`

Or maybe there's a layer in-between that I'm not yet aware of?

Reply:
> Or maybe there's a layer in-between that I'm not yet aware of?

Yeh, in `rhio` we are signing and encoding all messages in p2panda operations, with a new log per topic. This makes up the in-between layer. When operations arrive, we validate them against the local node store and then persist them (just in memory for now). 

But as you say, subscription-at-runtime is supported in `p2panda-net` and I expect we could make changes now to support it here as well. I think my early designs made it harder to do actually, probably now it's not such a problem.

The other aspect though, is that we are waiting for topics to be "ready" (another node joins the gossip overlay) before kicking off app messaging logic, so this would need looking at as well (although I don't think this "ready" flow is what we want in the end anyway, it's more a stop-gap until we have state sync).

Reply:
To me the rhio API has the benefit that you can wait for multiple topics to be ready before you can do anything.

That should have been possible before as well (see chat example) but maybe more verbose (but maybe also more clear?).

Another idea could be that .await is blocking on "topic_channel" until its ready. Or it returns a second future which _can_ be awaited (this is how iroh does it).

In any case, its good to keep in mind that all of this is temporary, this is why i thought the previous design (ready enum) was not perfect but fine for rhio: As soon as we have a persistence and sync layer we'll easily have writing to channels without being connected to any peers yet, the ready signal will mean again what it did originally (socket is ready) and we dont need to await anything.